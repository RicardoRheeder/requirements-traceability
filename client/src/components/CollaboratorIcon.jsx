import React from 'react'

const getRandColor = (brightness) => {
  // Six levels of brightness from 0 to 5, 0 being the darkest
  var rgb = [Math.random() * 256, Math.random() * 256, Math.random() * 256]
  var mix = [brightness * 51, brightness * 51, brightness * 51] //51 => 255/5
  var mixedrgb = [rgb[0] + mix[0], rgb[1] + mix[1], rgb[2] + mix[2]].map(
    function (x) {
      return Math.round(x / 2.0)
    }
  )
  return 'rgb(' + mixedrgb.join(',') + ')'
}

const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min) + min) //The maximum is exclusive and the minimum is inclusive
}

export default function CollaboratorIcon({ username }) {
  let randomColor = getRandColor(getRandomInt(3, 6))
  return (
    <div
      className="collaborator-icon"
      title={username}
      style={{
        backgroundImage:
          'radial-gradient(circle, ' +
          randomColor +
          ' 0%, ' +
          randomColor +
          ' 35%, #000000 100%)',
      }}
    >
      {username[0].toUpperCase()}
    </div>
  )
}
