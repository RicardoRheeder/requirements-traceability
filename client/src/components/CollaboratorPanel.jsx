import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { setUserColorObject } from '../redux/stores/common/actions'

import { CollaboratorIcon } from './'

const testUserObjects = [
  { userID: '12312', username: 'ricardo' },
  { userID: '54564', username: 'Faiz' },
  { userID: '84512', username: 'Devin' },
]

const getRandColor = (brightness) => {
  // Six levels of brightness from 0 to 5, 0 being the darkest
  var rgb = [Math.random() * 256, Math.random() * 256, Math.random() * 256]
  var mix = [brightness * 51, brightness * 51, brightness * 51] //51 => 255/5
  var mixedrgb = [rgb[0] + mix[0], rgb[1] + mix[1], rgb[2] + mix[2]].map(
    function (x) {
      return Math.round(x / 2.0)
    }
  )
  return 'rgb(' + mixedrgb.join(',') + ')'
}

const getRandomInt = (min, max) => {
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(Math.random() * (max - min) + min) //The maximum is exclusive and the minimum is inclusive
}

export default function CollaboratorPanel() {
  const dispatch = useDispatch()
  const userColorObject = useSelector((state) => state.common.userColorObject)

  const generateUserIcons = (userStruct) => {
    return userStruct.map(({ userID, username }, i) => {
      let randomColor

      if (userColorObject[username]) {
        randomColor = userColorObject[username]
      } else {
        randomColor = getRandColor(getRandomInt(3, 6))
        dispatch(setUserColorObject({[username]: randomColor}))
      }

      return (
        <CollaboratorIcon key={i} username={username} color={randomColor} />
      )
    })
  }

  return (
    <span className="collaborator-panel">
      {generateUserIcons(testUserObjects)}
    </span>
  )
}
