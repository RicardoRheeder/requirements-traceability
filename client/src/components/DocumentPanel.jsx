import React, { useEffect, useState } from 'react'
import Dropdown from 'react-dropdown'
import { useHistory } from 'react-router-dom'

import { useDispatch, useSelector } from 'react-redux'
import {
  setSelectedDocumentPanelObject,
  setModalObject,
  updateDataTree,
  setCurrentDocVersion,
} from '../redux/stores/common/actions'
import {
  getDocAsync,
  updateCurrentDocument,
} from '../redux/stores/document/actions'

export const DocumentPanel = ({ document }) => {
  const dispatch = useDispatch()
  const history = useHistory()

  const selectedDocumentPanelObject = useSelector(
    (state) => state.common.selectedDocumentPanelObject
  )

  const [selectedVersionTree, setSelectedVersionTree] = useState()
  const [versionList, setVersionList] = useState([])
  const [currentDropDownVersion, setCurrentDropDownVersion] = useState('')

  useEffect(() => {
    refreshVersionList()
  }, [])

  function refreshVersionList() {
    let defaultOption = '0.0'
    let tempVersionsList = []
    if (document.versions.length > 0) {
      // looping over versions array and parsing
      document.versions.forEach((version) => {
        const parsedVersion = JSON.parse(version)
        tempVersionsList.push(parsedVersion.versionName)
      })
      tempVersionsList.reverse()
      setSelectedVersionTree(JSON.parse(document.tree))
      // setting default option
      defaultOption = tempVersionsList[0]
      setCurrentDropDownVersion(defaultOption)
      setVersionList(tempVersionsList)
    } else {
      setCurrentDropDownVersion(defaultOption)
      setSelectedVersionTree(JSON.parse(document.tree))
    }
  }

  const _onDropdownSelect = (thing) => {
    // finding the corresponding tree for the version that was selected
    document.versions.forEach((version) => {
      const parsedVersion = JSON.parse(version)

      if (thing.value == parsedVersion.versionName) {
        dispatch(updateDataTree(JSON.parse(parsedVersion.tree)))
        setCurrentDropDownVersion(parsedVersion.versionName)
        setSelectedVersionTree(JSON.parse(parsedVersion.tree))
      }
    })
  }

  const inviteUserButton = () => {
    dispatch(setModalObject({ visible: true, mode: 2 }))
  }

  const openDocumentIntoEditor = () => {
    // fetching the current document
    dispatch(getDocAsync(document._id))
    // updating the selected version
    dispatch(updateDataTree(selectedVersionTree))
    // setting the current version of the document
    dispatch(setCurrentDocVersion(currentDropDownVersion))
    history.push('/editor')
  }

  return (
    <div
      className={
        'document-panel-component' +
        (selectedDocumentPanelObject._id == document._id ? ' selected' : '')
      }
      onClick={() => {
        dispatch(setSelectedDocumentPanelObject(document))
      }}
      onDoubleClick={openDocumentIntoEditor}
    >
      <div className="document-panel-title">
        <button className="add-person-button" onClick={inviteUserButton}>
          <img
            className="add-person-button-image"
            src="./assets/images/add-friend-icon.png"
          ></img>
        </button>
        <h2 className="doc-title">{document.title}</h2>
      </div>

      <div className="document-panel-dropdown">
        {/* <Dropdown
          options={versionList}
          onChange={_onDropdownSelect}
          value={currentDropDownVersion}
          placeholder="Select an option"
          className="dropdown-custom-wrapper"
          controlClassName="dropdown-custom-control"
          placeholderClassName="dropdown-custom-placeholder"
          menuClassName="dropdown-custom-menu"
          arrowClassName="dropdown-custom-arrow"
        /> */}
      </div>
    </div>
  )
}
