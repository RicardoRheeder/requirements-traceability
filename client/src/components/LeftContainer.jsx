import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { setModalObject } from '../redux/stores/common/actions'
import { useAuth0 } from '@auth0/auth0-react'
import { DocumentPanel } from './'
import { fetchUserDocsAsync } from '../redux/stores/document/actions'
import { fetchUserInfoAsync } from '../redux/stores/user/actions'
import { useState } from 'react'

export default function LeftContainer() {
  const { user } = useAuth0()
  const dispatch = useDispatch()
  const docs = useSelector((state) => state.document.documents)
  const selectedDoc = useSelector(
    (state) => state.common.selectedDocumentPanelObject
  )

  const [searchboxIsEmpty, setSearchboxIsEmpty] = useState(true)
  const [orderedDocList, setOrderedDocList] = useState([])

  // use effect to fetch the user info when the component mounts
  useEffect(() => {}, [docs])

  const updateSearch = (e) => {
    if (e.target.value.length == 0) {
      setSearchboxIsEmpty(true)
    } else {
      setSearchboxIsEmpty(false)
    }
    var matchingSearchs = []
    for (let i = 0; i < docs.length; i++) {
      var lowerCaseTitle = docs[i].title.toLowerCase()
      var lowerCaseSearchValue = e.target.value.toLowerCase()
      if (lowerCaseTitle.includes(lowerCaseSearchValue)) {
        console.log(docs[i].title + ' contains ' + e.target.value)
        matchingSearchs[matchingSearchs.length] = docs[i]
      }
    }
    setOrderedDocList(matchingSearchs)
  }

  const RenderDocumentPanels = (listOfDocs) => {
    console.log(listOfDocs)
    var orderedDocList = []
    for (let i = listOfDocs.length - 1; i >= 0; i--) {
      orderedDocList.push(listOfDocs[i])
    }
    if (orderedDocList.length != 0) {
      return orderedDocList.map((document, i) => {
        let tempDoc = JSON.parse(JSON.stringify(document))
        return <DocumentPanel document={tempDoc} key={i} />
      })
    }
    return (
      <div className="left-container-placeholder">
        <h2>No existing documents.</h2>
        <div>Click the 'Add' button below to get started.</div>
      </div>
    )
  }

  const addDocumentButton = () => {
    dispatch(setModalObject({ visible: true, mode: 0 }))
  }

  const removeDocumentButton = () => {
    dispatch(fetchUserInfoAsync(user))
    console.log(user);
    console.log(selectedDoc);
    var adminEmail = "";
    if(user.email==adminEmail){
      dispatch(setModalObject({ visible: true, mode: 1 }))
    }
  }

  return (
    <div className="left-container-root">
      <div className="document-searchbox-container">
        Search:{' '}
        <input
          className="document-searchbox"
          type="text"
          onChange={updateSearch}
        ></input>
      </div>
      <div className="display-area">
        {searchboxIsEmpty
          ? RenderDocumentPanels(docs)
          : RenderDocumentPanels(orderedDocList)}
      </div>
      <div className="add-remove-button-container">
        <button
          className="orange-button add-remove-button"
          onClick={addDocumentButton}
        >
          Add
        </button>

        <button
          className={
            'orange-button add-remove-button ' + (selectedDoc ? '' : 'disabled')
          }
          onClick={removeDocumentButton}
        >
          Remove
        </button>
      </div>
    </div>
  )
}
