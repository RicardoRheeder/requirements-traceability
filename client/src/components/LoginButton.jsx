import React from "react";

import { setLoggedIn } from "../redux/stores/common/actions";
import { useDispatch } from "react-redux";
import { useAuth0 } from "@auth0/auth0-react";

export default function LoginButton() {
  const dispatch = useDispatch();
  const { loginWithRedirect } = useAuth0();

  return (
    <button
      id="LogInButton"
      className="login-button"
      onClick={() => {
        // dispatch(setLoggedIn(true));
        loginWithRedirect();
      }}
    >
      Log in
    </button>
  );
}
