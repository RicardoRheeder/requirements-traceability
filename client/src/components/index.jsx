export { default as NavBar } from './NavBar'

export { default as Hierarchy } from './Editor/Hierarchy'

export { default as LoginButton } from './LoginButton'
export { default as LogoutButton } from './LogoutButton'

export { default as LeftContainer } from './LeftContainer'
export { DocumentPanel } from './DocumentPanel'

export { default as ModalComponent } from './Modal/ModalComponent'

<<<<<<< HEAD
export { default as RequirementStatusContainer } from "./RequirementStatus/RequirementStatusContainer"

=======
export { default as CollaboratorPanel } from './CollaboratorPanel'
export { default as CollaboratorIcon } from './CollaboratorIcon'
>>>>>>> master
