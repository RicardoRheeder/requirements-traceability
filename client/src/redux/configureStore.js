import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

// restricting extension in production using process.env.NODE_ENV which is set 'production'
import { composeWithDevTools } from 'redux-devtools-extension/logOnlyInProduction';

import store from './stores';

const enhancer = (process.env.NODE_ENV == 'development')
               ? composeWithDevTools(applyMiddleware(thunk))
               : applyMiddleware(thunk);

export default () => createStore(store, enhancer);
