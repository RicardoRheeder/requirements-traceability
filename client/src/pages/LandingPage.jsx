import React from 'react'
import { LoginButton } from '../components'

export default function LandingPage() {
  return (
    <div className="landing-page-root">
      <div className="logo-area">
        <img
          className="landing-logo-icon"
          src="./assets/images/Doc_Tracer_Banner_1_dark.png"
        ></img>
        <div>Let's get started!</div>
        <LoginButton />
      </div>
    </div>
  )
}
