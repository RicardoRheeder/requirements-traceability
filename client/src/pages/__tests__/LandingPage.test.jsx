import React from "react"
import Enzyme, { shallow, mount} from "enzyme"
import LandingPage from "../LandingPage"
import Adapter from "enzyme-adapter-react-16"
import { Provider } from "react-redux";
import configureStore from "redux-mock-store"
import { LoginButton } from "../../components";


const mockStore = configureStore([])
Enzyme.configure({adapter: new Adapter() });

describe("Landing Page", () => {
    let store;
    let landing;
    beforeEach(() => {
        store = mockStore({});
    });
    test("renders", () => {
        const wrapper = shallow(LandingPage());
        expect(wrapper.exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });
    test('renders with children', () => {
        const wrapper = mount(<Provider store = {store}><LandingPage/></Provider>);
        expect(wrapper.exists()).toBe(true);
        expect(wrapper).toMatchSnapshot();
    })
    test('renders button and logo', () => {
        const wrapper = mount(<Provider store = {store}><LandingPage/></Provider>);
        expect(wrapper.contains(<LoginButton/>)).toBe(true);
        expect(wrapper.find(".landing-logo-icon").exists()).toBe(true);
    })
});