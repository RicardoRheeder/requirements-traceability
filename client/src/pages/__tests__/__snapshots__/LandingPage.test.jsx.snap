// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing Page renders 1`] = `ShallowWrapper {}`;

exports[`Landing Page renders with children 1`] = `ReactWrapper {}`;
